# Data Wrangling

- `pandas`
- Where to find data?
   - Web Scraping & APIs
   
   
[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/COGS108/Lectures-Fa24/blob/main/03_pandas.ipynb)





<center>
<img src="https://raw.githubusercontent.com/COGS108/Lectures-Wi22/main/02_python/img/pandas.png" alt="pandas" width="600px">
</center>



Pandas is Python library for managing heterogenous data.

At its core, Pandas is used for the **DataFrame** object, which is:
- a data structure for labeled rows and columns of data
- associated methods and utilities for working with data.
- each column contains a `pandas` **Series**

# After this

When you are done with this exercise/lecture... the real learning can begin!

One of your best tools is https://pandastutor.com

Another is the documentation for pandas including https://pandas.pydata.org/docs/user_guide/10min.html

## Setup

In [1]:
# Import standard libraries
%matplotlib inline
import pandas as pd
import numpy as np

In [ ]:
# reminder about tab completion and contextual help

## Loading Data

In [3]:
# Load a csv file of data
df = pd.read_csv('data/my_data.csv')
print('total elements:',df.size,'\nshape of table',df.shape)

total elements: 1200 
shape of table (200, 6)


In [7]:
# Check out a few rows or last few rows of the dataframe using head() or tail()
df


id first_name last_name  age  score     value
0      295     Andrea     Clark   46     -1  24547.87
1      620       Bill     Woods   46    492  46713.90
2      891  Alexander  Jacobson   48    489  32071.74
3      914    Derrick   Bradley   52     -1  30650.48
4     1736    Allison    Thomas   44     -1   9553.12
..     ...        ...       ...  ...    ...       ...
195  97441     Krista     Ortiz   34     -1  24074.79
196  97728       Anna  Chambers   37    598      0.00
197  98115   Jennifer     Pitts   29    606   6876.75
198  98284   Brittany   Jenkins   34    665  43525.88
199  98366    Katelyn     Brown   45    501  29668.38

[200 rows x 6 columns]

Pandas DataFrame:
- Index for each row
- Column name for each column (Series)
- Stores heterogenous types

## Slicing

In [9]:
# Slicing (Indexing): select a Series (column) using its name

df['age']

0      46
1      46
2      48
3      52
4      44
       ..
195    34
196    37
197    29
198    34
199    45
Name: age, Length: 200, dtype: int64

In [12]:
df.loc[3:8, ['age', 'last_name'] ]

age last_name
3   52   Bradley
4   44    Thomas
5   57  Williams
6   46      Wood
7   50    Garcia
8   68   Coleman

In [14]:
df.iloc[4:8]

id first_name last_name  age  score      value
4  1736    Allison    Thomas   44     -1    9553.12
5  2049    Stephen  Williams   57    333  138936.92
6  2241      Malik      Wood   46     -1   10804.47
7  2607      Amber    Garcia   50    536    9367.27

In [ ]:
# Slicing: select a row & column with 'loc'
df.loc[4, 'age']

In [19]:
name_df = df.set_index('last_name')

name_df

In [20]:
name_df

id first_name  age  score     value
last_name                                        
Clark        295     Andrea   46     -1  24547.87
Woods        620       Bill   46    492  46713.90
Jacobson     891  Alexander   48    489  32071.74
Bradley      914    Derrick   52     -1  30650.48
Thomas      1736    Allison   44     -1   9553.12
...          ...        ...  ...    ...       ...
Ortiz      97441     Krista   34     -1  24074.79
Chambers   97728       Anna   37    598      0.00
Pitts      98115   Jennifer   29    606   6876.75
Jenkins    98284   Brittany   34    665  43525.88
Brown      98366    Katelyn   45    501  29668.38

[200 rows x 5 columns]

In [21]:
df.sort_values?

Signature:
df.sort_values(
    by: 'IndexLabel',
    *,
    axis: 'Axis' = 0,
    ascending: 'bool | list[bool] | tuple[bool, ...]' = True,
    inplace: 'bool' = False,
    kind: 'SortKind' = 'quicksort',
    na_position: 'str' = 'last',
    ignore_index: 'bool' = False,
    key: 'ValueKeyFunc | None' = None,
) -> 'DataFrame | None'
Docstring:
Sort by the values along either axis.

Parameters
----------
by : str or list of str
    Name or list of names to sort by.

    - if `axis` is 0 or `'index'` then `by` may contain index
      levels and/or column labels.
    - if `axis` is 1 or `'columns'` then `by` may contain column
      levels and/or index labels.
axis : "{0 or 'index', 1 or 'columns'}", default 0
     Axis to be sorted.
ascending : bool or list of bool, default True
     Sort ascending vs. descending. Specify list for multiple sort
     orders.  If this is a list of bools, must match the length of
     the by.
inplace : bool, default False
     If True, perform operation in-

In [25]:
df['age'] > 50


0      False
1      False
2      False
3       True
4      False
       ...  
195    False
196    False
197    False
198    False
199    False
Name: age, Length: 200, dtype: bool

#### Question #1

What would be the output of `df['age'] > 10`?

- A) subset of `df` including only rows of individuals older than 10
- B) a Boolean with `True` for rows where age is greater than 10 and `False` otherwise
- C) `id`s of rows where observations are greater than 10 
- D) an error
- E) I'm super lost

In [32]:
# to get subset
df.loc[ (df['age'] > 60) & (df['score']==-1) , ['first_name','last_name'] ]

first_name last_name
39     Kenneth   Johnson
158     Brenda       Cox

In [30]:
df.query('first_name == "Andrea" ')

id first_name last_name  age  score     value
0     295     Andrea     Clark   46     -1  24547.87
27  15137     Andrea   Simpson   55     -1  12155.47

In [ ]:
# NOTE: .loc uses the index, whatever it is
# sometimes the index is NOT numeric!
df.set_index('age').loc[52]

#set_index('first_name').loc['Andrea']


In [34]:
(
    df
    .set_index('first_name')
    .loc['Andr']
)

KeyError: 'Andr'

## Checking out the DataFrame

In [ ]:
# Check how large our dataframe is
df.shape

In [35]:
# Check what columns we have in our DataFrame
df.columns

Index(['id', 'first_name', 'last_name', 'age', 'score', 'value'], dtype='object')

In [ ]:
df.set_index('first_name').index

In [36]:
# Check the datatypes of our variables
df.dtypes

id              int64
first_name     object
last_name      object
age             int64
score           int64
value         float64
dtype: object

## Exploring the data

- quantitative (numbers)
- qualitative (categorical)
- basic descriptive statistics

In [38]:
# Checking categorical data
df['first_name'].value_counts().iloc[:20]

first_name
David        6
Michael      5
Eric         4
Charles      4
James        4
Elizabeth    3
Ashley       3
Jonathan     3
Sarah        3
Brian        3
Jennifer     3
Jason        3
John         3
Aaron        2
Sabrina      2
Robert       2
Lisa         2
Mary         2
Cynthia      2
Amanda       2
Name: count, dtype: int64

In [39]:
# Check a particular descriptive statistic, like mean(), median(), mode(), std(), var(), etc
df[['id','age', 'value']].median()

id       54643.50
age         46.00
value    17976.51
dtype: float64

In [40]:
# Describe a particular column
df['value'].describe()

count       189.000000
mean      28730.336296
std       32493.945741
min           0.000000
25%        9593.030000
50%       17976.510000
75%       33163.310000
max      204999.960000
Name: value, dtype: float64

In [41]:
# Get descriptive statistics of all numerical columns
df.describe()

id         age       score          value
count    200.000000  200.000000  200.000000     189.000000
mean   52929.150000   46.020000  416.595000   28730.336296
std    29414.298899   10.028582  237.176674   32493.945741
min      295.000000   14.000000   -1.000000       0.000000
25%    26709.500000   39.000000  288.750000    9593.030000
50%    54643.500000   46.000000  463.500000   17976.510000
75%    80840.750000   53.000000  596.500000   33163.310000
max    98366.000000   69.000000  942.000000  204999.960000

#### Question #2

What's the average (mean) age of the individuals in this dataset?

- A) 14
- B) 46
- C) 28730
- D) NA
- E) I'm super lost/unsure

In [ ]:
# there are two commands that could give you the correct answer...
df['age'].mean()

## `pandas`: Common Manipulations

You'll want to be *very* familiar with a few common data manipulations when wrangling data, each of which is described below:

Manipulation | Description
-------|------------
**select** | select which columns to include in dataset
**filter** | filter dataset to only include specified rows
**mutate** | add a new column based on values in other columns
**groupby** | group values to apply a function within the specified groups
**summarize** | calculate specified summary metric of a specified variable
**arrange** | sort rows ascending or descending order of a specified column
**merge** | join separate datasets into a single dataset based on a common column



In [ ]:
import this

## Selecting & Dropping Columns

- include subset of columns of larger data frame

In [ ]:
df.head()

In [ ]:
# specify which columns to include
select_df = df[['id', 'age', 'score', 'value']]
select_df.head()

In [ ]:
# Drop rows we don't want
new_df = df.drop(labels=[0,2,6], axis='rows')

In [ ]:
# Check out the DataFrame after dropping, what went wrong and how do we fix it?
new_df.head()

## Filtering Data (slicing)

- include a subset (slice) of rows from larger data frame

In [ ]:
# Check if we have any data from people below the age of 18
sum(df['age'] < 18)

In [ ]:
df.loc[(df['age'] < 18)]

In [ ]:
# Select only participants who are 18 or older  AND who have a score of -1
df_new = df[ (df['age'] >= 18)] # & (df['score']==-1)]
df_new

## Missing Data (NaNs)

In [ ]:
# Check for missing values
df['value'].hasn

In [ ]:
df.isnull?

In [ ]:
# note in class
# can operate on entire dataframe
df.isnull()

In [ ]:
# Check for null values by row
df.isnull().sum(axis='rows')

In [ ]:
# Have a look at the missing values
df[df['value'].isnull()]
# can also use .isna() or isnan()!

## Dealing with Missing Data - NaNs

In [ ]:
# Dealing with null values: Drop rows with missing data
print('before dropping null rows',df.shape)
no_na_df = df.dropna()
print('new shape', no_na_df.shape)

In [ ]:
# or you can fill in a value for those missing values!
df.fillna?
# df.fillna(100) or df.fillna( df['value'].mean() ) or just df.fillna()

## Finding Missing Data - Bad Values

In [ ]:
# Check for the properties of specific columns
df['score'].describe()

In [ ]:
df[df.score<0].shape

In [ ]:
# Check the plot of the data for score to see the distribution
df['score'].plot(kind='hist', bins=25);

In [ ]:
df.hist();

## Dealing with Missing Data - Bad Values

In [ ]:
# Look for how many values have a -1 value in 'score'
sum(df['score'] == -1)

In [ ]:
# Drop any row with -1 value in 'score'
df = df[df['score'] != -1]
df.shape

## Creating new columns (mutating)

- `assign` can be very helpful in adding a new column
- lambda functions can be used to carry out calculations

In [ ]:
# convert age in years to age in (approximate) days
df = df.assign(age_days = df['age'] * 365)
df.head()

In [ ]:
df['age_months'] = df['age'] * 12
df.head()

## Grouping & summarizing

- group by a particular variable
- calculate summary statistics/metrics within group

In [ ]:
df.first_name.value_counts()

In [ ]:
# caclculate average within each age
df.groupby('first_name').age.mean()

## Sorting Rows (arrange)

- specify order in which to display rows

In [ ]:
df.head()

In [ ]:
# sort by values in age
df = df.sort_values(by = ['age'],ascending=True)
df.head()

## Combining datasets
![image of join operations](https://raw.githubusercontent.com/COGS108/Lectures-Fa24/main/img/join.png)

In [ ]:
## Create two DataFrames
left = pd.DataFrame({'key': ['A', 'B', 'C', 'D'], 'value': np.random.randn(4)})    
right = pd.DataFrame({'key': ['B', 'D', 'E', 'F'], 'value': np.random.randn(4)})

In [ ]:
left

In [ ]:
right

In [ ]:
left.merge?

In [ ]:
left.merge(right, on='key', )

In [ ]:
# inner merge by default
pd.merge(left, right, on='key')

In [ ]:
# same as above
pd.merge(left, right, on='key', how='inner')

In [ ]:
# right merge
pd.merge(left, right, on='key', how='right')

In [ ]:
# left merge
pd.merge(left, right, on='key', how='left')

In [ ]:
# outer join
pd.merge(left, right, on='key', how='outer')

In [ ]:
pd.merge(left, right, on='key', how='outer').mean()

Let's do this as a check in! 
## https://forms.gle/EpNS7BYhfAxvBM4X7

#### Question #3

If table A had 5 rows and table B had 5 rows and 3 of those rows in each table were from the same observations present in the *other* table, how many rows would be present if an **inner merge** were carried out?

- A) 3
- B) 5
- C) 10
- D) 13
- E) Totally unsure

#### Question #4

If table A had 5 rows and table B had 5 rows and 3 of those rows in each table were from the same observations present in the *other* table, how many rows would be present if a **left merge** were carried out?

- A) 3
- B) 5
- C) 10
- D) 13
- E) Totally unsure

## Visualizing Data

- We'll have a whole lecture (or two) on visualization
- For now, we'll just look at one uniquely-pandas approach

In [ ]:
# Plot all numerical columns, and their interactions
pd.plotting.scatter_matrix(df[['age', 'score', 'value']], figsize=[12, 12], marker=12);  

## Application Program Interface (APIs)

- APIs are basically a way for software to talk to software 
    - They are an interface into an application / website / database designed for computers / software.

Notes on APIs:
- Follow API guidelines! 
- These guidelines typically specify the number / rate / size of requests

## Github API

You can access the github api with the following API. Just added specifiers for what you are looking for. 

https://api.github.com/

For example, the following URL will search for the user 'ShanEllis':

https://api.github.com/users/shanellis




## Requesting Web Pages from Python

In [ ]:
# The requests module allows you to send URL requests from python
import requests  
from bs4 import BeautifulSoup

In [ ]:
# Request data from the Github API on a particular user
page = requests.get('https://api.github.com/users/jasongfleischer')  

In [ ]:
# The content we get back is a messily organized json file
page.content

#### Question #5

What type/format of output is this?

- A) CSV
- B) XML
- C) JSON
- D) API
- E) I'm super lost

In [ ]:
# We can read in the json data with pandas
git_data = pd.read_json(page.content, typ='series')

In [ ]:
# Check out the pandas series object full of data
git_data  

### Authorized Access - OAuth

Open Authorization is a protocol to authorize access (of a user / application) to an API.

OAuth provides a secure way to 'log-in' without using account names and passwords. 

It is effectively a set of keys, and passwords you can use to access APIs. 

## Web Scraping vs. APIs

Web scraping and APIs are different approaches:

- APIs are an interface to interact with an application, designed for programmatic use
    - They allow systematic, controlled access to (for example) and applications database
    - They typically return structured (friendly) data 

- Web scraping (typically) involves navigating through the internet, programmatically following an architecture built for humans
    - This can be hard to systematize, being dependent on the idiosyncracies of a web page, at the time you request it
    - This typically returns relatively unstructured data
    - This entails much more wrangling of the data

## Where to Find Data?

* [Awesome Public Datasets](https://github.com/awesomedata/awesome-public-datasets/blob/master/README.rst)
* [Data.gov](https://catalog.data.gov/dataset)
* [Data Is Plural](https://docs.google.com/spreadsheets/d/1wZhPLMCHKJvwOkP4juclhjFgqIY8fQFMemwKL2c64vk/edit#gid=0)
* [UCSD Datasets](https://ucsd.libguides.com/data-statistics/home)
* [Datasets | Deep Learning](http://deeplearning.net/datasets/)
* [Stanford | Social Science Data Collection](https://data.stanford.edu/)
* [Eviction Lab (email required)](https://evictionlab.org/get-the-data/)
* [San Diego Data](https://data.sandiego.gov/)
* [US Census](https://www.census.gov/)
* [Open Climate Data](http://openclimatedata.net/)
* [Data and Story Library](https://dasl.datadescription.com/datafiles/)
* [UCSD behavioral mobile data](http://extrasensory.ucsd.edu/)
* [Kaggle](https://www.kaggle.com/)
* [FiveThirtyEight](https://data.fivethirtyeight.com/)
* [data.world](https://data.world/)
* [Free Datasets - R and Data Mining ](http://www.rdatamining.com/resources/data)
* [Data Sources for Cool Data Science Projects](https://blog.thedataincubator.com/2014/10/data-sources-for-cool-data-science-projects-part-1/)

## Notes on Working with Data

### Data Science is Ad-Hoc

- It is part of the job description to put things together that were not designed to go together.
- We do not have universal solutions, but haphazard, idiosyncratic systems, for data collection, storage and analysis.
- Data is everywhere. But relatively little of it was collected *as data*.

### Data Collection, Curation, and Storage are Difficult

- It can be difficult to choose broadly useful standards
- Take time to think about your data, and how you will load, store, organize and save it

### Data is Inherently Noisy

- We live in a messy, noisy, world, with messy, noisy, people, using messy, noisy instruments.
- There is no perfect data. 
    - There is better / or worse data, given the context.

### Different Objectives

- Humans and computers are different.
- We interact with '*data*' in different ways.
- This underlies many aspects of data wrangling
    - The 'friendliness' of data types / files
    - The difference between web scraping and APIs
    - A disconnect between data in the real world, and data we want to use

## So... What to do?

- Think about how your data are stored & its structure?
- Look at your data before you anayze it
    - are there missing values? 
    - outlier values? 
- Are your data trustworthy? 
    - source?
    - how was it generated?

## Specific Recommendations

- Prioritize using well structured, common, open file types
    - Take advantage of existing tools to deal with these files (numpy, pandas, etc.)

- Look into, and then follow, common conventions
    - Minimize custom objects, workflows and data files 
- Look for APIs. Ask if they are available.
    - Acknowledge that web scraping and/or wrangling unstructured data are complex / long tasks

- Think about data flow from the beginning. Organize your data pipeline, consider the 'wrangling' aspects throughout
    - Set yourself up with well organized, labelled approach to your data
    - Think about when and how you might want/need to save out intermediate results.